File: android/app/src/main/java/com/desklink/android/data/security/PairingAuth.kt
package com.desklink.android.data.security

import com.desklink.android.domain.model.ProtocolConstants
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

// The client/server contexts bind each HMAC proof to its direction so a proof cannot be replayed the other way. Golden vectors: tools/protocol_vectors.py.
object PairingAuth {

    fun clientProof(key: ByteArray, serverNonce: ByteArray, clientNonce: ByteArray): ByteArray =
        proof(ProtocolConstants.AUTH_CLIENT_CONTEXT, key, serverNonce, clientNonce)

    fun serverProof(key: ByteArray, serverNonce: ByteArray, clientNonce: ByteArray): ByteArray =
        proof(ProtocolConstants.AUTH_SERVER_CONTEXT, key, serverNonce, clientNonce)

    // Constant-time comparison to avoid leaking proof bytes via timing.
    fun verify(received: ByteArray, expected: ByteArray): Boolean =
        MessageDigest.isEqual(received, expected)

    private fun proof(
        context: String,
        key: ByteArray,
        serverNonce: ByteArray,
        clientNonce: ByteArray,
    ): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        mac.update(context.toByteArray(Charsets.UTF_8))
        mac.update(serverNonce)
        mac.update(clientNonce)
        return mac.doFinal()
    }
}

File: macos/DeskLink/Sources/Data/Security/PairingAuth.swift
import CryptoKit
import Foundation

// Distinct client/server contexts bind each proof to its direction (no cross-replay); must match Android byte-for-byte, golden vectors in tools/protocol_vectors.py.
enum PairingAuth {

    static func clientProof(key: Data, serverNonce: Data, clientNonce: Data) -> Data {
        proof(context: ProtocolConstants.authClientContext, key: key, serverNonce: serverNonce, clientNonce: clientNonce)
    }

    static func serverProof(key: Data, serverNonce: Data, clientNonce: Data) -> Data {
        proof(context: ProtocolConstants.authServerContext, key: key, serverNonce: serverNonce, clientNonce: clientNonce)
    }

    // Constant-time comparison; do not replace with == (timing side-channel).
    static func verify(_ received: Data, expected: Data) -> Bool {
        guard received.count == expected.count else { return false }
        var difference: UInt8 = 0
        for (a, b) in zip(received, expected) { difference |= a ^ b }
        return difference == 0
    }

    private static func proof(context: String, key: Data, serverNonce: Data, clientNonce: Data) -> Data {
        var message = Data(context.utf8)
        message.append(serverNonce)
        message.append(clientNonce)
        let code = HMAC<SHA256>.authenticationCode(for: message, using: SymmetricKey(data: key))
        return Data(code)
    }
}
